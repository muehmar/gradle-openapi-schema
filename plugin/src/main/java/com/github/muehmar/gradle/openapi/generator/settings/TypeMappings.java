package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import lombok.Value;

@Value
public class TypeMappings {
  PList<ClassTypeMapping> classTypeMappings;
  PList<FormatTypeMapping> formatTypeMappings;
  PList<DtoMapping> dtoMappings;

  public static TypeMappings empty() {
    return new TypeMappings(PList.empty(), PList.empty(), PList.empty());
  }

  public static TypeMappings ofSingleClassTypeMapping(ClassTypeMapping classTypeMapping) {
    return new TypeMappings(PList.single(classTypeMapping), PList.empty(), PList.empty());
  }

  public static TypeMappings ofClassTypeMappings(ClassTypeMapping... classTypeMapping) {
    return new TypeMappings(PList.fromArray(classTypeMapping), PList.empty(), PList.empty());
  }

  public static TypeMappings ofSingleFormatTypeMapping(FormatTypeMapping formatTypeMapping) {
    return new TypeMappings(PList.empty(), PList.single(formatTypeMapping), PList.empty());
  }

  public static TypeMappings ofSingleDtoMapping(DtoMapping dtoMapping) {
    return new TypeMappings(PList.empty(), PList.empty(), PList.single(dtoMapping));
  }
}
