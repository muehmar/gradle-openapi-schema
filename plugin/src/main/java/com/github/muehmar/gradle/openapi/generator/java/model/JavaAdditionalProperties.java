package com.github.muehmar.gradle.openapi.generator.java.model;

import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType.javaAnyType;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaName MAP_PROPERTY_NAME = JavaName.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, javaAnyType());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, javaAnyType());
  }

  public static JavaName additionalPropertiesName() {
    return MAP_PROPERTY_NAME;
  }

  public boolean isNotAllowed() {
    return not(allowed);
  }

  public boolean isValueAnyType() {
    return type.equals(javaAnyType());
  }

  public boolean isNotValueAnyType() {
    return not(type.equals(javaAnyType()));
  }

  public TechnicalPojoMember asTechnicalPojoMember() {
    return TechnicalPojoMember.additionalProperties(type);
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(type.getQualifiedClassName().getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
