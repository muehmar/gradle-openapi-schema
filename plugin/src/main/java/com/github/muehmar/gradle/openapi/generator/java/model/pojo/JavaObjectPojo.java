package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder.fullMemberContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder.fullWitherContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ConstructorContentBuilder.fullConstructorContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsContentBuilder.fullEqualsContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeContentBuilder.fullHashCodeContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringContentBuilder.fullToStringContentBuilder;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.AnyOfContainer;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.OneOfContainer;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.name.SchemaName;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.codegenerator.java.JavaModifier;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final SchemaName schemaName;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      SchemaName schemaName,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.schemaName = schemaName;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.fromPojoName(type.mapName(objectPojo.getName().getPojoName())),
        objectPojo.getName().getSchemaName(),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(
                comp ->
                    JavaOneOfComposition.wrap(
                        comp, objectPojo.getDiscriminator(), type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(schemaName.asName());
  }

  @Override
  public JavaName getClassName() {
    return name.asJavaName();
  }

  public JavaPojoName getJavaPojoName() {
    return name;
  }

  public JavaName prefixedClassNameForMethod(String prefix) {
    return name.asJavaName().prefixedMethodName(prefix);
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  /** This includes also any possible required additional property. */
  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size() + requiredAdditionalProperties.size();
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    final PList<JavaPojoMember> allOfMembers =
        allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> oneOfMembers =
        oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> anyOfMembers =
        anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
    return allOfMembers.concat(oneOfMembers).concat(anyOfMembers).distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    return getAllMembers()
        .flatMap(JavaPojoMember::getTechnicalMembers)
        .add(additionalProperties.asTechnicalPojoMember());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public PList<JavaObjectPojo> getOneOfPojos() {
    return oneOfComposition
        .map(JavaOneOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasOneOfComposition() {
    return oneOfComposition.isPresent();
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaObjectPojo> getAnyOfPojos() {
    return anyOfComposition
        .map(JavaAnyOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasAnyOfComposition() {
    return anyOfComposition.isPresent();
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return fullMemberContentBuilder()
        .isArrayPojo(false)
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return fullHashCodeContentBuilder().technicalPojoMembers(getTechnicalMembers()).build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return fullEqualsContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return fullToStringContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return fullConstructorContentBuilder()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .modifier(Optional.of(JavaModifier.PUBLIC).filter(ignore -> not(isSimpleMapPojo())))
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return fullWitherContentBuilder()
        .className(getClassName())
        .membersForWithers(getMembers())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public Optional<OneOfContainer> getOneOfContainer() {
    return oneOfComposition.map(composition -> new OneOfContainer(name, composition));
  }

  public Optional<AnyOfContainer> getAnyOfContainer() {
    return anyOfComposition.map(composition -> new AnyOfContainer(name, composition));
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
