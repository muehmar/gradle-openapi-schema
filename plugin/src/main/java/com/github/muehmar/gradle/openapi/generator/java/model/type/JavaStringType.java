package com.github.muehmar.gradle.openapi.generator.java.model.type;

import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NOT_NULLABLE;

import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassNames;
import com.github.muehmar.gradle.openapi.generator.java.model.type.api.ApiType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.api.TypeMapping;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.EnumMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode(callSuper = true)
@ToString
public class JavaStringType extends NonGenericJavaType {
  private static final Map<StringType.Format, QualifiedClassName> FORMAT_CLASS_NAME_MAP =
      createFormatClassNameMap();

  private final Constraints constraints;

  private static Map<StringType.Format, QualifiedClassName> createFormatClassNameMap() {
    final Map<StringType.Format, QualifiedClassName> map = new EnumMap<>(StringType.Format.class);
    map.put(StringType.Format.DATE, QualifiedClassNames.LOCAL_DATE);
    map.put(StringType.Format.DATE_TIME, QualifiedClassNames.LOCAL_DATE_TIME);
    map.put(StringType.Format.TIME, QualifiedClassNames.LOCAL_TIME);
    map.put(StringType.Format.BINARY, QualifiedClassNames.BYTE_ARRAY);
    map.put(StringType.Format.UUID, QualifiedClassNames.UUID);
    map.put(StringType.Format.URL, QualifiedClassNames.URL);
    map.put(StringType.Format.URI, QualifiedClassNames.URI);
    return map;
  }

  protected JavaStringType(
      QualifiedClassName className,
      Optional<ApiType> apiType,
      Nullability nullability,
      Constraints constraints) {
    super(className, apiType, nullability);
    this.constraints = constraints;
  }

  public static JavaStringType wrap(StringType stringType, TypeMappings typeMappings) {
    final QualifiedClassName originalClassName = internalClassNameFromFormat(stringType);
    final TypeMapping typeMapping = mapType(stringType, typeMappings, originalClassName);
    return new JavaStringType(
        typeMapping.getClassName(),
        typeMapping.getApiType(),
        stringType.getNullability(),
        stringType.getConstraints());
  }

  private static TypeMapping mapType(
      StringType stringType, TypeMappings typeMappings, QualifiedClassName internalClassName) {

    final TypeMapping formatTypeMapping =
        TypeMapping.fromFormatMappings(
            internalClassName, stringType.getFormatString(), typeMappings.getFormatTypeMappings());

    final TypeMapping classTypeMapping =
        TypeMapping.fromClassMappings(internalClassName, typeMappings.getClassTypeMappings());

    return formatTypeMapping.or(classTypeMapping, internalClassName);
  }

  public static JavaStringType noFormat() {
    return new JavaStringType(
        QualifiedClassNames.STRING, Optional.empty(), NOT_NULLABLE, Constraints.empty());
  }

  private static QualifiedClassName internalClassNameFromFormat(StringType stringType) {
    return Optional.ofNullable(FORMAT_CLASS_NAME_MAP.get(stringType.getFormat()))
        .orElse(QualifiedClassNames.STRING);
  }

  @Override
  public boolean isJavaArray() {
    return className.equals(QualifiedClassNames.BYTE_ARRAY);
  }

  @Override
  public Constraints getConstraints() {
    return constraints;
  }

  @Override
  public JavaType withNullability(Nullability nullability) {
    return new JavaStringType(className, apiType, nullability, constraints);
  }

  @Override
  public <T> T fold(
      Function<JavaArrayType, T> onArrayType,
      Function<JavaBooleanType, T> onBooleanType,
      Function<JavaEnumType, T> onEnumType,
      Function<JavaMapType, T> onMapType,
      Function<JavaAnyType, T> onAnyType,
      Function<JavaNumericType, T> onNumericType,
      Function<JavaIntegerType, T> onIntegerType,
      Function<JavaObjectType, T> onObjectType,
      Function<JavaStringType, T> onStringType) {
    return onStringType.apply(this);
  }
}
