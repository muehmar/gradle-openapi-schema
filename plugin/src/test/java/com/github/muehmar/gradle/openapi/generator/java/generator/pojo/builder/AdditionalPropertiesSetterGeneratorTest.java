package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.builder;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.builder.AdditionalPropertiesSetterGenerator.additionalPropertiesSetterGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.requiredBirthdate;
import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes.objectType;
import static com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMappings.STRING_MAPPING_WITH_CONVERSION;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaStringType;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class AdditionalPropertiesSetterGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("objectPojoAnyTypeAdditionalProperties")
  void generate_when_objectPojoAnyTypeAdditionalProperties_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = additionalPropertiesSetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(requiredBirthdate()), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("allNecessityAndNullabilityVariantsPojo")
  void generate_when_allNecessityAndNullabilityVariantsPojo_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = additionalPropertiesSetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.allNecessityAndNullabilityVariants(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("objectPojoSpecificTypeAdditionalProperties")
  void generate_when_objectPojoSpecificTypeAdditionalProperties_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = additionalPropertiesSetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(
                PList.single(requiredBirthdate()),
                JavaAdditionalProperties.allowedFor(objectType())),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("stringTypeWithMappingAdditionalProperties")
  void generate_when_stringTypeWithMappingAdditionalProperties_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = additionalPropertiesSetterGenerator();

    final TypeMappings stringTypeMapping =
        TypeMappings.ofClassTypeMappings(STRING_MAPPING_WITH_CONVERSION);
    final JavaStringType stringType = JavaStringType.wrap(StringType.noFormat(), stringTypeMapping);

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(
                PList.single(requiredBirthdate()), JavaAdditionalProperties.allowedFor(stringType)),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("nullableStringTypeWithMappingAdditionalProperties")
  void generate_when_nullableStringTypeWithMappingAdditionalProperties_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = additionalPropertiesSetterGenerator();

    final TypeMappings stringTypeMapping =
        TypeMappings.ofClassTypeMappings(STRING_MAPPING_WITH_CONVERSION);
    final JavaStringType stringType =
        JavaStringType.wrap(
            StringType.noFormat().withNullability(Nullability.NULLABLE), stringTypeMapping);

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(
                PList.single(requiredBirthdate()), JavaAdditionalProperties.allowedFor(stringType)),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
