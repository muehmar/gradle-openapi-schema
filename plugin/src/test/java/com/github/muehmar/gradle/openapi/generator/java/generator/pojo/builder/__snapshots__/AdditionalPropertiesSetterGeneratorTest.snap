allNecessityAndNullabilityVariantsPojo=[
com.fasterxml.jackson.annotation.JsonAnySetter
com.github.muehmar.openapi.util.Tristate
java.util.HashMap
java.util.Map

@JsonAnySetter
public Builder addAdditionalProperty(String key, Object value) {
  this.additionalProperties.put(key, value);
  return this;
}

public Builder addAdditionalProperty(String key, Tristate<Object> value) {
  value
      .onValue(val -> this.additionalProperties.put(key, val))
      .onNull(() -> this.additionalProperties.put(key, null))
      .onAbsent(() -> null);
  return this;
}

public Builder setAdditionalProperties(Map<String, Object> additionalProperties) {
  this.additionalProperties = new HashMap<>(additionalProperties);
  return this;
}
]


nullableStringTypeWithMappingAdditionalProperties=[
com.custom.CustomString
com.fasterxml.jackson.annotation.JsonAnySetter
com.github.muehmar.openapi.util.Tristate
java.util.HashMap
java.util.Map

@JsonAnySetter
public Builder addAdditionalProperty(String key, CustomString value) {
  this.additionalProperties.put(key, value != null ? CustomString.toString(value) : null);
  return this;
}

public Builder addAdditionalProperty(String key, Tristate<CustomString> value) {
  value
      .onValue(val -> this.additionalProperties.put(key, CustomString.toString(val)))
      .onNull(() -> this.additionalProperties.put(key, null))
      .onAbsent(() -> null);
  return this;
}

public Builder setAdditionalProperties(Map<String, CustomString> additionalProperties) {
  this.additionalProperties = new HashMap<>();
  additionalProperties.forEach((key, val) -> this.additionalProperties.put(key, val != null ? CustomString.toString(val) : null));
  return this;
}
]


objectPojoAnyTypeAdditionalProperties=[
com.fasterxml.jackson.annotation.JsonAnySetter
com.github.muehmar.openapi.util.Tristate
java.util.HashMap
java.util.Map

@JsonAnySetter
public Builder addAdditionalProperty(String key, Object value) {
  this.additionalProperties.put(key, value);
  return this;
}

public Builder addAdditionalProperty(String key, Tristate<Object> value) {
  value
      .onValue(val -> this.additionalProperties.put(key, val))
      .onNull(() -> this.additionalProperties.put(key, null))
      .onAbsent(() -> null);
  return this;
}

public Builder setAdditionalProperties(Map<String, Object> additionalProperties) {
  this.additionalProperties = new HashMap<>(additionalProperties);
  return this;
}
]


objectPojoSpecificTypeAdditionalProperties=[
com.fasterxml.jackson.annotation.JsonAnySetter
java.util.HashMap
java.util.Map
java.util.Optional

@JsonAnySetter
public Builder addAdditionalProperty(String key, UserDto value) {
  this.additionalProperties.put(key, value);
  return this;
}

public Builder addAdditionalProperty(String key, Optional<UserDto> value) {
  value.ifPresent(val -> this.additionalProperties.put(key, val));
  return this;
}

public Builder setAdditionalProperties(Map<String, UserDto> additionalProperties) {
  this.additionalProperties = new HashMap<>(additionalProperties);
  return this;
}
]


stringTypeWithMappingAdditionalProperties=[
com.custom.CustomString
com.fasterxml.jackson.annotation.JsonAnySetter
java.util.HashMap
java.util.Map
java.util.Optional

@JsonAnySetter
public Builder addAdditionalProperty(String key, CustomString value) {
  this.additionalProperties.put(key, value != null ? CustomString.toString(value) : null);
  return this;
}

public Builder addAdditionalProperty(String key, Optional<CustomString> value) {
  value.ifPresent(val -> this.additionalProperties.put(key, CustomString.toString(val)));
  return this;
}

public Builder setAdditionalProperties(Map<String, CustomString> additionalProperties) {
  this.additionalProperties = new HashMap<>();
  additionalProperties.forEach((key, val) -> this.additionalProperties.put(key, val != null ? CustomString.toString(val) : null));
  return this;
}
]